src/targets/gpu/include/migraphx/gpu/context.hpp

```diff
@@ -84,32 +86,93 @@ struct hip_device
         shared<rocblas_handle_ptr> rbhandle = nullptr;
     };
 
-    void add_stream() { streams.emplace_back(device_id); }
+    static hip_event_ptr create_event()
+    {
+        hipEvent_t event;
+        auto status = hipEventCreateWithFlags(&event, hipEventDisableTiming);
+        if(status != hipSuccess)
+            MIGRAPHX_THROW("Failed to creat event");
+        return hip_event_ptr{event};
+    }
+
+    void add_streams()
+    {
+        int num_of_streams = 1;
+        assert(streams.empty());
+        if(enabled(MIGRAPHX_DISABLE_NULL_STREAM{}))
+            num_of_streams = stream_info().num_of_streams();
+        for(int i = 0; i < num_of_streams; ++i)
+            streams.emplace_back(device_id);
+    }
 
     stream& get_stream() { return streams.at(current_stream); }
 
     void set_stream(std::size_t n) { current_stream = n; }
+    void create_events(int num_of_events)
+    {
+        for(int i = events.size(); i < num_of_events; ++i)
+            events.emplace_back(create_event());
+    }
+    void record_event(int event)
+    {
+        create_events(event + 1);
+        hipEventRecord(events.at(event).get(), streams.at(current_stream).get());
+    }
+
+    void wait_event(int event)
+    {
+        hipStreamWaitEvent(streams.at(current_stream).get(), events.at(event).get(), 0);
+    }
+
+    void stream_sync()
+    {
+        if(enabled(MIGRAPHX_DISABLE_NULL_STREAM{}))
+        {
+            int num_of_streams = streams.size();
+            if(num_of_streams > 0)
+            {
+                for(int i = 0; i < num_of_streams; i++)
+                    hipStreamSynchronize(streams.at(i).get());
+            }
+        }
+    }
 
     private:
     std::size_t device_id      = 0;
     std::size_t current_stream = 0;
     std::vector<stream> streams;
+    std::vector<shared<hip_event_ptr>> events;
 };
 
 struct context
 {
     context(std::size_t n = 0) : current_device(std::make_shared<hip_device>(n)) {}
 
-    hip_device& get_current_device()
+    hip_device& get_current_device() const
```

Comment:
This should return `const hip_device&`. Actually, there should be two overloads one for `const` that returns `const hip_device&` and another for non-const that returns `hip_device&`.

Reply:
Overload is not allowed if function diffs only in return types.